src/ui/public/state_management/unhash_states.js

```diff
@@ -0,0 +1,46 @@
+import { parse as parseUrl, format as formatUrl } from 'url';
+import { mapValues } from 'lodash';
+
+export function UnhashStatesProvider(getAppState, globalState) {
+  const getDefaultStates = () => [getAppState(), globalState].filter(Boolean);
+
+  this.inAbsUrl = (url, states = getDefaultStates()) => {
+    if (!url) return url;
+
+    const entireUrlParsed = parseUrl(url, true);
+    if (!entireUrlParsed.hostname) {
+      // passing a url like "localhost:5601" or "/app/kibana" should be prevented
+      throw new TypeError(
+        'Only absolute urls should be passed to `unhashStates.inAbsUrl()`. ' +
+        'Unable to detect url hostname.'
+      );
+    }
+
+    if (!entireUrlParsed.hash) return url;
+
+    const appUrl = entireUrlParsed.hash.slice(1); // trim the #
+    if (!appUrl) return url;
+
+    const appUrlParsed = parseUrl(entireUrlParsed.hash.slice(1), true);
+    if (!appUrlParsed.query) return url;
+
+    const unhashedAppQuery = this.inObject(appUrlParsed.query || {}, states);
+    return formatUrl({
+      protocol: entireUrlParsed.protocol,
+      host: entireUrlParsed.host,
+      pathname: entireUrlParsed.pathname,
+      query: entireUrlParsed.query,
+      hash: formatUrl({
+        pathname: appUrlParsed.pathname,
+        query: unhashedAppQuery,
+      })
+    });
+  };
+
+  this.inObject = (query, states = getDefaultStates()) => {
+    return mapValues(query, (val, key) => {
+      const state = states.find(s => key === s.getQueryParamName());
+      return state ? state.toRISON() : val;
+    });
+  };
+}
```

Comment:
I was having a hard time understanding the role of this module and what it was doing internally, so I tried renaming things in a way that told me a story. Do you think these changes below make the logic easier to follow?

Also, I think it's possible to split each of these methods into their own individual modules. It doesn't seem necessary to group them into one. I mostly came to that conclusion because I was having a hard time figuring out a better name for this file/module, and then I realized it was because these methods don't seem cohesive enough.

```javascript
/**
 * get_url_with_states.js
 * Convert a URL that contains hash references to one that contains RISON-encoded values.
 */
export default function getUrlWithStates(urlWithHashes) => {
  if (!urlWithHashes) return urlWithHashes;

  const urlWithHahesObject = parseUrl(urlWithHashes, true);
  if (!urlWithHahesObject.hostname) {
    // passing a url like "localhost:5601" or "/app/kibana" should be prevented
    throw new TypeError(
      'Only absolute urls should be passed to `unhashStates.inAbsUrl()`. ' +
      'Unable to detect url hostname.'
    );
  }

  if (!urlWithHahesObject.hash) return urlWithHashes;

  // e.g. /discover?_g=h@44136&_a=h@cdf09
  const clientRoute = urlWithHahesObject.hash.slice(1); // trim the #
  if (!clientRoute) return urlWithHashes;

  const clientRouteObject = parseUrl(clientRoute, true);

  // e.g. {_g: h@44136, _a: h@cdf09}
  const queryParamToHashMap = clientRouteObject.query;
  if (!queryParamToHashMap) return urlWithHashes;

  // e.g. {_g: (rison), _a: (rison)}
  const states = [getAppState(), globalState].filter(Boolean);
  const queryParamToStateMap = mapQueryParamsToStateValues(queryParamToHashMap, states);

  const urlWithStates = formatUrl({
    protocol: urlObject.protocol,
    host: urlObject.host,
    pathname: urlObject.pathname,
    query: urlObject.query,
    hash: formatUrl({
      pathname: clientRouteObject.pathname,
      query: queryParamToStateMap,
    })
  });

  return urlWithStates;
};
```
```javascript
/**
 * map_query_params_to_state_values.js
 * Map query-param-keyed hash references to one that contains RISON-encoded tate values.
 */
export default function mapQueryParamsToStateValues(queryParamToHashMap, states) {
  return mapValues(queryParamToHashMap, (val, key) => {
    // Pull out the state for each query param in the query.
    const state = states.find(s => key === s.getQueryParamName());
    return state ? state.toRISON() : val;
  });
};
```